apiserver/params/internal.go

```diff
@@ -342,6 +342,13 @@ type ControllerAPIInfoResults struct {
 	Results []ControllerAPIInfoResult `json:"results"`
 }
 
+// RelationApplicationUnit holds a relation, an application and a unit tag.
+type RelationApplicationUnit struct {
+	Relation    string `json:"relation"`
+	Application string `json:"application"`
```

Comment:
Is it possible to have a application that is different from the unit?  Do we want this to be possible?  If not, should we derive the application tag from the unit tag?  names.UnitApplication.

Reply:
Hmmm... I think you are right. I just added it there for future proofing (just in case we ever decide to allow `Application != unitApplication(unit)`) given what happened with the unit-that-is-really-an-app-tag field in `RelationUnit`. If you think it's better to omit it, I can actually drop the new struct and replace it with `RelationUnit`.

Reply:
Agreed, overloading a tag is bad.  Perhaps we just need to be very clear in expectations between the 2 tags.  This would primarily be an issue if used from pylibjuju I'm thinking.  

If the app from the unit tag doesn't match the app, it errors? Otherwise results from "u.leadershipChecker.LeadershipCheck(appTag.Id(), unitTag.Id())" could be interesting.

Reply:
I think having the new "ReadLocalApplicationSettings" is the crux of the change needed.
And having that new API means that RelationUnit struct is all that's needed.
The check would be that the unit attribute either matches the auth tag (if a unit is connecting) or is a valid unit of the application (if an application is connecting).

Reply:
The PR currently includes the auth'd unit matches the unit in the params struct. Is the latter check for the k8s case actually required given that you are only allowed to proceed if you are the leader? A bogus unit name would always fail the leadership check and get back an ErrPerm. Or do we want to be more explicit about the check?